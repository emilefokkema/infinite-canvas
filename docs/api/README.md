ef-infinite-canvas

# ef-infinite-canvas

## Enumerations

- [Units](api/enums/Units.md)

## Interfaces

- [Config](api/interfaces/Config.md)
- [DrawEvent](api/interfaces/DrawEvent.md)
- [EventMap](api/interfaces/EventMap.md)
- [InfiniteCanvas](api/interfaces/InfiniteCanvas.md)
- [InfiniteCanvasCtr](api/interfaces/InfiniteCanvasCtr.md)
- [InfiniteCanvasEventHandlers](api/interfaces/InfiniteCanvasEventHandlers.md)
- [InfiniteCanvasEventWithDefaultBehavior](api/interfaces/InfiniteCanvasEventWithDefaultBehavior.md)
- [InfiniteCanvasRenderingContext2D](api/interfaces/InfiniteCanvasRenderingContext2D.md)
- [InfiniteCanvasTouch](api/interfaces/InfiniteCanvasTouch.md)
- [InfiniteCanvasTouchEvent](api/interfaces/InfiniteCanvasTouchEvent.md)
- [InfiniteCanvasTouchList](api/interfaces/InfiniteCanvasTouchList.md)
- [TransformationEvent](api/interfaces/TransformationEvent.md)
- [TransformationRepresentation](api/interfaces/TransformationRepresentation.md)
- [Transformed](api/interfaces/Transformed.md)

## Variables

### default

â€¢ **default**: [`InfiniteCanvasCtr`](api/interfaces/InfiniteCanvasCtr.md)

#### Defined in

[infinite-canvas.ts:49](https://github.com/emilefokkema/infinite-canvas/blob/4a1afe1/src/api-surface/infinite-canvas.ts#L49)
