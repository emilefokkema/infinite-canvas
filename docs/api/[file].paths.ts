import { loadPaths } from '../.vitepress/api-docs/load-paths'

export default {
    async paths(){
        if(process.env.NODE_ENV !== 'production'){
            return [];
        }
        return await loadPaths();
    }
}