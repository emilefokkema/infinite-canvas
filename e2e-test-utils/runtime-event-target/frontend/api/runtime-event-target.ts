import { ChartMap } from '../../shared/serializable-types'

export interface RuntimeEventTarget<TMap>{
    emitEvents(map: ChartMap<TMap>): void
    handleEvents<TType extends keyof TMap>(type: TType, handler: (e: TMap[TType]) => void): void
    switchToCapture(type: keyof TMap): void
}