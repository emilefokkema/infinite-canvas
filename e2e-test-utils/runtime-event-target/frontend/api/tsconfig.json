{
    "compilerOptions": {
        "module": "ESNext",
        "moduleResolution": "bundler",
        "target": "ESNext",
        "composite": true,
        "emitDeclarationOnly": true,
        "outDir": "./dist",
        "lib": ["DOM"]
    },
    "references": [
        {"path": "../../shared"}
    ],
    "include": ["**/*"]
}