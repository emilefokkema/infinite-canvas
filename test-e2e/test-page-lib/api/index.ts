import type { InfiniteCanvas } from '../../../src/api-surface/infinite-canvas'
import { CanvasElementInitialization, InfiniteCanvasInitialization } from './configuration';
import { CanvasElementWrapper, AttachedEventListener } from './wrappers';
import { EventTarget } from './event-target'

declare global{
    interface Window{
        TestPageLib: TestPageLib
        TestCaseLib: TestCaseLib
    }
}

export const EVENT_LISTENER_DATA = 'EVENT_LISTENER_DATA'

export interface TestPageLib{
    initializeCanvasElement(config: CanvasElementInitialization): CanvasElementWrapper
    initializeInfiniteCanvas(config: CanvasElementInitialization & InfiniteCanvasInitialization): Promise<InfiniteCanvas>
    openMessagePort(url: string): Promise<void>;
    addEventListener<
        TEventName,
        TEventType,
        TEventTarget extends EventTarget<TEventType, TEventName>
        >(target: TEventTarget, name: TEventName, capture: boolean | undefined, eventListenerId: string): AttachedEventListener<TEventType>
    makeSerializableTextMetrics(textMetrics: TextMetrics): TextMetrics
}

export interface TestCaseLib{
    drawTestCase(testCaseFile: string): Promise<void>
}

export type { CanvasElementInitialization, InfiniteCanvasInitialization, CanvasElementWrapper, EventTarget, AttachedEventListener }