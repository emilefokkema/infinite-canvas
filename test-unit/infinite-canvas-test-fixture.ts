import { Mock, MockInstance, vi } from 'vitest'
import { DrawingLock } from "src/drawing-lock";
import { MockCanvasMeasurementProvider } from './mock-canvas-measurement-provider';
import { CanvasContextMock } from './canvas-context-mock';
import { Config } from 'api/config';
import { MockDrawingIterationProvider } from './mock-drawing-iteration-provider';
import { CssLengthConverterFactory } from 'src/css-length-converter-factory';
import { InfiniteCanvasViewBox } from 'src/infinite-canvas-viewbox';
import { RectangleManagerImpl } from 'src/rectangle/rectangle-manager-impl';
import { InfiniteContext } from 'src/infinite-context/infinite-context';
import { ViewBox } from 'src/interfaces/viewbox';
import { InfiniteCanvasRenderingContext2D } from 'api/infinite-canvas-rendering-context-2d';


function setupGlobals(){
	window.createImageBitmap = function(): Promise<ImageBitmap>{return undefined;};
	(<any>window).ImageData = class {
		public width: number;
		public height: number;
		public data: Uint8ClampedArray;
		constructor(arrayOrWidth: Uint8ClampedArray | number, widthOrHeight: number, height?: number){
			if(typeof arrayOrWidth === "number"){
				this.width = arrayOrWidth;
				this.height = widthOrHeight;
			}else{
				this.width = widthOrHeight;
				this.height = height;
				this.data = arrayOrWidth;
			}
		}
	};
}

export interface InfiniteCanvasTestFixture{
    width: number
    height: number
    releaseDrawingLockSpy: MockInstance<()=>void>
    getDrawingLockSpy: MockInstance<()=>DrawingLock>
    measurementProvider: MockCanvasMeasurementProvider
    contextMock: CanvasContextMock
    config: Partial<Config>
    mockDrawingIterationProvider: MockDrawingIterationProvider
    viewbox: ViewBox,
    infiniteContext: InfiniteCanvasRenderingContext2D
}

export function createInfiniteCanvasTestFixture(): InfiniteCanvasTestFixture{
    setupGlobals();
    const drawingLock: DrawingLock = {release(){}};
    const releaseDrawingLockSpy = vi.spyOn(drawingLock, 'release');
    const getDrawingLockSpy = vi.fn<()=>DrawingLock>().mockReturnValue(drawingLock)
    const width = 200;
    const height = 200;
    const measurementProvider = new MockCanvasMeasurementProvider(width, height);
    const contextMock = new CanvasContextMock();
    const config = {}
    const mockDrawingIterationProvider = new MockDrawingIterationProvider()
    const cssLengthConverterFactory: CssLengthConverterFactory = {
        create: () => ({
            getNumberOfPixels(value: number){
                return value;
            }
        })
    };
    const viewbox = new InfiniteCanvasViewBox(
        new RectangleManagerImpl(measurementProvider, config),
        contextMock.mock,
        mockDrawingIterationProvider,
        getDrawingLockSpy,
        () => false);
    const infiniteContext = new InfiniteContext(undefined, viewbox, cssLengthConverterFactory);
    return {
        width,
        height,
        releaseDrawingLockSpy,
        getDrawingLockSpy,
        measurementProvider,
        contextMock,
        config,
        mockDrawingIterationProvider,
        viewbox,
        infiniteContext
    };
}