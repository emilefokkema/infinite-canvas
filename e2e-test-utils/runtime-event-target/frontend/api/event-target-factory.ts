import { EventTargetLike } from '../../shared/event-target-like'
import { ConnectionData } from '../../shared/connection-data';
import { RuntimeEventTarget } from './runtime-event-target'

export interface EventTargetFactory{
    createEventTarget<TMap>(
        eventTargetLike: EventTargetLike<TMap>,
        connection: ConnectionData
    ): RuntimeEventTarget<TMap>
}