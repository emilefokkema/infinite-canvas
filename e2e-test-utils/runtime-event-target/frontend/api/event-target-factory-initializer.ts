import { EventTargetFactory } from "./event-target-factory";

export type EventTargetFactoryInitializer = () => EventTargetFactory