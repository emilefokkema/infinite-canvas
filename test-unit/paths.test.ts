import { beforeEach, describe, expect, it } from 'vitest'
import {CanvasContextMock} from "./canvas-context-mock";
import {InfiniteCanvasRenderingContext2D} from "api/infinite-canvas-rendering-context-2d";
import { createInfiniteCanvasTestFixture } from './infinite-canvas-test-fixture';

describe('an infinite canvas context', () => {
	let infiniteContext: InfiniteCanvasRenderingContext2D;
	let contextMock: CanvasContextMock;

	beforeEach(() => {
		({ contextMock, infiniteContext} = createInfiniteCanvasTestFixture());
	});

    describe("that makes a path, fills it and then fills an overlapping rect", () => {

		beforeEach(() => {
			infiniteContext.fillStyle = "#f00";
			infiniteContext.beginPath();
			infiniteContext.moveTo(0, 0);
			infiniteContext.lineTo(2, 0);
			infiniteContext.lineTo(2, 2);
			infiniteContext.lineTo(0, 2);
			infiniteContext.lineTo(0, 0);
			infiniteContext.fill();
			infiniteContext.fillStyle = "#00f";
			contextMock.clear();
			infiniteContext.fillRect(1, 1, 2, 2);
		});

		it("should draw the two rectangles in the right order", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe("and then fills the path again", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.fill();
			});

			it("should not have forgotten the previous path", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe("that draws a path, fills it and then adds to the path", () => {

		beforeEach(() => {
			infiniteContext.fillStyle = "#f00";
			infiniteContext.beginPath();
			infiniteContext.moveTo(30,30);
			infiniteContext.lineTo(30,100);
			infiniteContext.lineTo(100,100);
			infiniteContext.fill();
			infiniteContext.lineTo(100,30);
		});

		describe("and then strokes the path", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.stroke();
			});

			it("should have executed the last path modification only once", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe("that fills a rectangle, creates a path inside it, clears a rectangle inside the first rectangle and then fills the created path", () => {

		beforeEach(() => {
			infiniteContext.fillStyle = "#f00";
			infiniteContext.fillRect(0, 0, 100, 100);
			infiniteContext.fillStyle = "#00f";
			infiniteContext.beginPath();
			infiniteContext.moveTo(50,0);
			infiniteContext.lineTo(50,50);
			infiniteContext.lineTo(0,50);
			infiniteContext.lineTo(0, 0);
			infiniteContext.closePath();
			infiniteContext.clearRect(0, 0, 75, 75);
			contextMock.clear();
			infiniteContext.fill();
		});

		it("should have executed everything", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that makes a path, fills it, clears it completely with a clearRect, expands it and strokes it", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.moveTo(1, 1);
			infiniteContext.lineTo(1, 2);
			infiniteContext.lineTo(2, 2);
			infiniteContext.fill();
			infiniteContext.clearRect(0, 0, 3, 3);
			infiniteContext.lineTo(2, 1);
			contextMock.clear();
			infiniteContext.stroke();
		});

		it("should no longer fill the path", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that creates a rectangular path, fills another rectangle and then fills the created path", () => {

		beforeEach(() => {
			infiniteContext.fillStyle = "#f00";
			infiniteContext.beginPath();
			infiniteContext.rect(0,0,50,50);
			infiniteContext.fillRect(50,50,50,50);
			contextMock.clear();
			infiniteContext.fill();
		});

		it("should have filled both rectangles", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that creates a rectangular path, strokes another rectangle and then strokes the created path", () => {

		beforeEach(() => {
			infiniteContext.strokeStyle = "#f00";
			infiniteContext.beginPath();
			infiniteContext.rect(0,0,50,50);
			infiniteContext.strokeRect(50,50,50,50);
			contextMock.clear();
			infiniteContext.stroke();
		});

		it("should have stroked both rectangles", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that draws a line without calling 'moveTo'", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.lineTo(10, 10);
			infiniteContext.lineTo(20, 10);
			contextMock.clear();
			infiniteContext.stroke();
		});

		it("should draw a line", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    
})