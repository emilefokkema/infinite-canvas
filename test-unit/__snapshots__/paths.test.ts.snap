// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`an infinite canvas context > that creates a rectangular path, fills another rectangle and then fills the created path > should have filled both rectangles 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(50,50)",
  "context.lineTo(100,50)",
  "context.lineTo(100,100)",
  "context.lineTo(50,100)",
  "context.closePath()",
  "context.moveTo(50,50)",
  "context.fill()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(50,0)",
  "context.lineTo(50,50)",
  "context.lineTo(0,50)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that creates a rectangular path, strokes another rectangle and then strokes the created path > should have stroked both rectangles 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.strokeStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(50,50)",
  "context.lineTo(100,50)",
  "context.lineTo(100,100)",
  "context.lineTo(50,100)",
  "context.closePath()",
  "context.moveTo(50,50)",
  "context.stroke()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(50,0)",
  "context.lineTo(50,50)",
  "context.lineTo(0,50)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a line without calling 'moveTo' > should draw a line 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(20,10)",
  "context.lineWidth = 1",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a path, fills it and then adds to the path > and then strokes the path > should have executed the last path modification only once 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(30,30)",
  "context.lineTo(30,100)",
  "context.lineTo(100,100)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(30,30)",
  "context.lineTo(30,100)",
  "context.lineTo(100,100)",
  "context.lineWidth = 1",
  "context.lineTo(100,30)",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that fills a rectangle, creates a path inside it, clears a rectangle inside the first rectangle and then fills the created path > should have executed everything 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(100,0)",
  "context.lineTo(100,100)",
  "context.lineTo(0,100)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0,0,75,75)",
  "context.restore()",
  "context.beginPath()",
  "context.moveTo(50,0)",
  "context.lineTo(50,50)",
  "context.lineTo(0,50)",
  "context.lineTo(0,0)",
  "context.closePath()",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that makes a path, fills it and then fills an overlapping rect > and then fills the path again > should not have forgotten the previous path 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.lineTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.lineTo(0,0)",
  "context.fillStyle = "#00f"",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that makes a path, fills it and then fills an overlapping rect > should draw the two rectangles in the right order 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.lineTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that makes a path, fills it, clears it completely with a clearRect, expands it and strokes it > should no longer fill the path 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(1,2)",
  "context.lineTo(2,2)",
  "context.lineWidth = 1",
  "context.lineTo(2,1)",
  "context.stroke()",
]
`;
